demo_keyframe_bundle_adjustment_meta/apps/main_program.cpp

```diff
@@ -0,0 +1,178 @@
+    lidar_data_file.read(reinterpret_cast<char*>(&lidar_data_buffer[0]), num_elements*sizeof(float));
+    return lidar_data_buffer;
+}
+
+int main(int argc, char** argv)
+{
+    ros::init(argc, argv, "kitti_helper");
+    ros::NodeHandle n("~");
+    std::string dataset_folder, sequence_number, output_bag_file;
+    n.getParam("dataset_folder", dataset_folder);
+    n.getParam("sequence_number", sequence_number);
+    std::cout << "Reading sequence " << sequence_number << " from " << dataset_folder << '\n';
+    bool to_bag;
+    n.getParam("to_bag", to_bag);
+    if (to_bag)
+        n.getParam("output_bag_file", output_bag_file);
+    int publish_delay;
+    n.getParam("publish_delay", publish_delay);
+    publish_delay = publish_delay <= 0 ? 1 : publish_delay;
+
+    ros::Publisher pub_laser_cloud = n.advertise<sensor_msgs::PointCloud2>("/kitti/velo/pointcloud", 2);
+
+    image_transport::ImageTransport it(n);
+    image_transport::Publisher pub_image_left = it.advertise("/image_left", 2);
+    image_transport::Publisher pub_image_right = it.advertise("/image_right", 2);
+
+    ros::Publisher pubOdomGT = n.advertise<nav_msgs::Odometry> ("/odometry_gt", 5);
+    nav_msgs::Odometry odomGT;
+    odomGT.header.frame_id = "/camera_init";
+    odomGT.child_frame_id = "/ground_truth";
+
+    ros::Publisher pubPathGT = n.advertise<nav_msgs::Path> ("/path_gt", 5);
+    nav_msgs::Path pathGT;
+    pathGT.header.frame_id = "/camera_init";
+
+    std::string timestamp_path = "sequences/" + sequence_number + "/times.txt";
+    std::ifstream timestamp_file(dataset_folder + timestamp_path, std::ifstream::in);
+
+    std::string ground_truth_path = "results/" + sequence_number + ".txt";
+    std::ifstream ground_truth_file(dataset_folder + ground_truth_path, std::ifstream::in);
+
+    rosbag::Bag bag_out;
+    if (to_bag)
+        bag_out.open(output_bag_file, rosbag::bagmode::Write);
+    
+    Eigen::Matrix3d R_transform;
+    R_transform << 0, 0, 1, -1, 0, 0, 0, -1, 0;
+    Eigen::Quaterniond q_transform(R_transform);
+
+    std::string line;
+    std::size_t line_num = 0;
+
+    ros::Rate r(10.0 / publish_delay);
+    while (std::getline(timestamp_file, line) && ros::ok())
+    {
+        float timestamp = stof(line);
+        std::stringstream left_image_path, right_image_path;
+        left_image_path << dataset_folder << "sequences/" + sequence_number + "/image_0/" << std::setfill('0') << std::setw(6) << line_num << ".png";
+        cv::Mat left_image = cv::imread(left_image_path.str(), CV_LOAD_IMAGE_GRAYSCALE);
+        right_image_path << dataset_folder << "sequences/" + sequence_number + "/image_1/" << std::setfill('0') << std::setw(6) << line_num << ".png";
+        cv::Mat right_image = cv::imread(left_image_path.str(), CV_LOAD_IMAGE_GRAYSCALE);
+
+        std::getline(ground_truth_file, line);
+        std::stringstream pose_stream(line);
+        std::string s;
+        Eigen::Matrix<double, 3, 4> gt_pose;
```

Comment:
I personally would prefer an Eigen::Affine3d but thats more a matter of taste I guess

Reply:
I will change it to Affine3d Johannes